--- ./Main/Forms/VolumePasswordWizardPage.h	2012-02-07 14:36:48.000000000 +0400
+++ ../veracrypt_1.17_Source/src/Main/Forms/VolumePasswordWizardPage.h	2016-02-14 12:11:21.000000000 +0300
@@ -1,9 +1,13 @@
 /*
- Copyright (c) 2008 TrueCrypt Developers Association. All rights reserved.
+ Derived from source code of TrueCrypt 7.1a, which is
+ Copyright (c) 2008-2012 TrueCrypt Developers Association and which is governed
+ by the TrueCrypt License 3.0.
 
- Governed by the TrueCrypt License 3.0 the full text of which is contained in
- the file License.txt included in TrueCrypt binary and source code distribution
- packages.
+ Modifications and additions to the original source code (contained in this file) 
+ and all other portions of this file are Copyright (c) 2013-2016 IDRIX
+ and are governed by the Apache License 2.0 the full text of which is
+ contained in the file License.txt included in VeraCrypt binary and source
+ code distribution packages.
 */
 
 #ifndef TC_HEADER_Main_Forms_VolumePasswordWizardPage
@@ -12,7 +16,7 @@
 #include "Forms.h"
 #include "VolumePasswordPanel.h"
 
-namespace TrueCrypt
+namespace VeraCrypt
 {
 	class VolumePasswordWizardPage : public VolumePasswordWizardPageBase
 	{
@@ -22,6 +26,11 @@
 
 		shared_ptr <KeyfileList> GetKeyfiles () const { return PasswordPanel->GetKeyfiles(); }
 		shared_ptr <VolumePassword> GetPassword () const { return PasswordPanel->GetPassword(); }
+		void EnableUsePim () { PasswordPanel->EnableUsePim (); }
+		bool IsPimSelected () const { return PasswordPanel->IsUsePimChecked ();}
+		void SetPimSelected (bool selected) const { PasswordPanel->SetUsePimChecked (selected);}
+		
+		shared_ptr <Pkcs5Kdf> GetPkcs5Kdf () const { bool bUnsupportedKdf; return PasswordPanel->GetPkcs5Kdf(bUnsupportedKdf); }
 		bool IsValid ();
 		void SetMaxStaticTextWidth (int width) { InfoStaticText->Wrap (width); }
 		void SetPageText (const wxString &text) { InfoStaticText->SetLabel (text); }
