--- ./Common/Common.rc	2012-02-07 14:36:48.000000000 +0400
+++ ../veracrypt_1.17_Source/src/Common/Common.rc	2016-02-14 12:11:21.000000000 +0300
@@ -28,31 +28,31 @@
 
 IDD_ABOUT_DLG DIALOGEX 31, 51, 292, 199
 STYLE DS_SETFONT | DS_MODALFRAME | DS_3DLOOK | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
-CAPTION "About TrueCrypt"
-CLASS "SplashDlg"
+CAPTION "About VeraCrypt"
+CLASS "VeraCryptSplashDlg"
 FONT 8, "MS Shell Dlg", 0, 0, 0x0
 BEGIN
     EDITTEXT        IDC_ABOUT_CREDITS,7,111,277,45,ES_MULTILINE | WS_VSCROLL | NOT WS_TABSTOP
     DEFPUSHBUTTON   "OK",IDOK,230,178,52,14
     LTEXT           "",IDC_HOMEPAGE,18,87,117,9,SS_NOTIFY
     LTEXT           "",IDT_ABOUT_RELEASE,18,71,235,8
-    CONTROL         517,IDC_ABOUT_BKG,"Static",SS_BITMAP,0,0,12,11,WS_EX_STATICEDGE
+    CONTROL         IDB_TEXTUAL_LOGO_BKG,IDC_ABOUT_BKG,"Static",SS_BITMAP,0,0,12,11,WS_EX_STATICEDGE
     LTEXT           "",IDT_ABOUT_VERSION,18,61,161,8
     CONTROL         "",IDC_STATIC,"Static",SS_ETCHEDHORZ,1,167,291,1,WS_EX_STATICEDGE
     CONTROL         "",IDC_STATIC,"Static",SS_ETCHEDHORZ,1,169,291,1,WS_EX_STATICEDGE
     CONTROL         "",IDC_STATIC,"Static",SS_ETCHEDHORZ,1,107,291,1,WS_EX_STATICEDGE
     CONTROL         "",IDC_ABOUT_LOGO_AREA,"Static",SS_GRAYRECT | NOT WS_VISIBLE,0,0,293,50,WS_EX_TRANSPARENT | WS_EX_STATICEDGE
-    CONTROL         518,IDC_TEXTUAL_LOGO_IMG,"Static",SS_BITMAP,12,26,157,16
+    CONTROL         IDB_TEXTUAL_LOGO_96DPI,IDC_TEXTUAL_LOGO_IMG,"Static",SS_BITMAP,12,26,157,16
 END
 
-IDD_COMMANDHELP_DLG DIALOGEX 0, 0, 249, 213
+IDD_COMMANDHELP_DLG DIALOGEX 0, 0, 249, 274
 STYLE DS_SETFONT | DS_MODALFRAME | DS_3DLOOK | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
 CAPTION "Command Line Help"
-CLASS "CustomDlg"
+CLASS "VeraCryptCustomDlg"
 FONT 8, "MS Shell Dlg", 0, 0, 0x0
 BEGIN
-    DEFPUSHBUTTON   "OK",IDOK,93,191,59,14
-    LTEXT           "",IDC_COMMANDHELP_TEXT,20,11,208,174
+    DEFPUSHBUTTON   "OK",IDOK,93,252,59,14
+    LTEXT           "",IDC_COMMANDHELP_TEXT,20,10,208,241
 END
 
 IDD_RAWDEVICES_DLG DIALOGEX 0, 0, 305, 209
@@ -65,34 +65,42 @@
     PUSHBUTTON      "Cancel",IDCANCEL,248,190,50,14
 END
 
-IDD_MOUNT_OPTIONS DIALOGEX 0, 0, 277, 172
+IDD_MOUNT_OPTIONS DIALOGEX 0, 0, 277, 231
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
-CAPTION "TrueCrypt - Mount Options"
+CAPTION "VeraCrypt - Mount Options"
 FONT 8, "MS Shell Dlg", 400, 0, 0x1
 BEGIN
     CONTROL         "Mount volume as read-&only",IDC_MOUNT_READONLY,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,12,11,194,10
     CONTROL         "Mount volume as removable &medium",IDC_MOUNT_REMOVABLE,
                     "Button",BS_AUTOCHECKBOX | WS_TABSTOP,12,25,195,10
+    CONTROL         "Use backup header embedded in &volume if available",IDC_USE_EMBEDDED_HEADER_BAK,
+                    "Button",BS_AUTOCHECKBOX | WS_TABSTOP,12,39,257,11
     CONTROL         "Mount partition &using system encryption without pre-boot authentication",IDC_MOUNT_SYSENC_PART_WITHOUT_PBA,
                     "Button",BS_AUTOCHECKBOX | WS_TABSTOP,12,53,259,11
+    EDITTEXT        IDC_VOLUME_LABEL,112,67,150,14,ES_AUTOHSCROLL
     CONTROL         "&Protect hidden volume against damage caused by writing to outer volume",IDC_PROTECT_HIDDEN_VOL,
-                    "Button",BS_AUTOCHECKBOX | WS_TABSTOP,16,86,252,10
-    EDITTEXT        IDC_PASSWORD_PROT_HIDVOL,112,104,151,14,ES_PASSWORD | ES_AUTOHSCROLL
-    CONTROL         "&Display password",IDC_SHOW_PASSWORD_MO,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,112,123,90,10
-    CONTROL         "U&se keyfiles",IDC_KEYFILES_ENABLE_HIDVOL_PROT,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,112,136,90,10
-    PUSHBUTTON      "&Keyfiles...",IDC_KEYFILES_HIDVOL_PROT,203,125,60,14
-    LTEXT           "What is hidden volume protection?",IDC_LINK_HIDVOL_PROTECTION_INFO,16,151,247,10,SS_NOTIFY
+                    "Button",BS_AUTOCHECKBOX | WS_TABSTOP,16,97,252,10
+    EDITTEXT        IDC_PASSWORD_PROT_HIDVOL,112,115,151,14,ES_PASSWORD | ES_AUTOHSCROLL
+    COMBOBOX        IDC_PKCS5_PRF_ID,112,136,91,90,CBS_DROPDOWNLIST | WS_TABSTOP
+    EDITTEXT        IDC_PIM,112,156,42,14,ES_RIGHT | ES_PASSWORD | ES_AUTOHSCROLL | ES_NUMBER | NOT WS_VISIBLE
+    CONTROL         "Use P&IM",IDC_PIM_ENABLE,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,112,161,97,10
+    LTEXT           "(Empty or 0 for default iterations)",IDC_PIM_HELP,158,159,112,8,NOT WS_VISIBLE
+    CONTROL         "&Display password",IDC_SHOW_PASSWORD_MO,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,112,174,90,10
+    CONTROL         "U&se keyfiles",IDC_KEYFILES_ENABLE_HIDVOL_PROT,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,112,187,90,10
+    PUSHBUTTON      "&Keyfiles...",IDC_KEYFILES_HIDVOL_PROT,204,183,60,14
     DEFPUSHBUTTON   "OK",IDOK,211,7,60,14
     PUSHBUTTON      "Cancel",IDCANCEL,211,24,60,14
-    RTEXT           "P&assword to hidden volume:\n(if empty, cache is used)",IDT_HIDDEN_PROT_PASSWD,15,103,91,17,0,WS_EX_RIGHT
-    GROUPBOX        "Hidden Volume Protection",IDT_HIDDEN_VOL_PROTECTION,6,72,265,95
-    CONTROL         "Use backup header embedded in &volume if available",IDC_USE_EMBEDDED_HEADER_BAK,
-                    "Button",BS_AUTOCHECKBOX | WS_TABSTOP,12,39,257,11
+    LTEXT           "What is hidden volume protection?",IDC_LINK_HIDVOL_PROTECTION_INFO,16,202,247,10,SS_NOTIFY
+    RTEXT           "P&assword to hidden volume:\n(if empty, cache is used)",IDT_HIDDEN_PROT_PASSWD,15,114,91,17,0,WS_EX_RIGHT
+    GROUPBOX        "Hidden Volume Protection",IDT_HIDDEN_VOL_PROTECTION,6,83,265,136
+    RTEXT           "PKCS-5 PRF:",IDT_PKCS5_PRF,15,137,91,17
+    RTEXT           "Volume PIM:",IDT_PIM,15,159,91,17,NOT WS_VISIBLE
+    LTEXT           "Volume Label in Windows:",IDT_VOLUME_LABEL,12,70,95,8
 END
 
-IDD_KEYFILES DIALOGEX 0, 0, 345, 237
+IDD_KEYFILES DIALOGEX 0, 0, 345, 251
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
-CAPTION "TrueCrypt - Keyfiles"
+CAPTION "VeraCrypt - Keyfiles"
 FONT 8, "MS Shell Dlg", 400, 0, 0x1
 BEGIN
     CONTROL         "",IDC_KEYLIST,"SysListView32",LVS_REPORT | LVS_SHOWSELALWAYS | LVS_ALIGNLEFT | LVS_NOSORTHEADER | WS_BORDER | WS_TABSTOP,7,8,263,118
@@ -101,20 +109,22 @@
     PUSHBUTTON      "Add &Token Files...",IDC_TOKEN_FILES_ADD,139,132,65,14
     PUSHBUTTON      "&Remove",IDC_KEYREMOVE,209,132,61,14
     PUSHBUTTON      "Remove &All",IDC_KEYREMOVEALL,275,132,61,14
-    CONTROL         "U&se keyfiles",IDC_KEYFILES_ENABLE,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,7,219,83,11
-    PUSHBUTTON      "&Generate Random Keyfile...",IDC_GENERATE_KEYFILE,213,217,123,14
+    CONTROL         "U&se keyfiles",IDC_KEYFILES_ENABLE,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,7,232,83,11
+    PUSHBUTTON      "&Generate Random Keyfile...",IDC_GENERATE_KEYFILE,213,230,123,14
     DEFPUSHBUTTON   "OK",IDOK,279,8,59,14
     PUSHBUTTON      "Cancel",IDCANCEL,279,25,59,14
     LTEXT           "",IDT_KEYFILES_NOTE,10,161,324,41,0,WS_EX_TRANSPARENT
     LTEXT           "WARNING: If you lose a keyfile or if any bit of its first 1024 kilobytes changes, it will be impossible to mount volumes that use the keyfile!",IDT_KEYFILE_WARNING,279,44,58,85,0,WS_EX_TRANSPARENT
     CONTROL         "",IDC_STATIC,"Static",SS_ETCHEDHORZ,2,154,343,1,WS_EX_STATICEDGE
     CONTROL         "",IDC_STATIC,"Static",SS_ETCHEDHORZ,2,209,343,1,WS_EX_STATICEDGE
-    LTEXT           "More information on keyfiles",IDC_LINK_KEYFILES_INFO,96,220,108,10,SS_NOTIFY
+    LTEXT           "More information on keyfiles",IDC_LINK_KEYFILES_INFO,96,233,108,10,SS_NOTIFY
+    CONTROL         "Try first to mount with an empty password",IDC_KEYFILES_TRY_EMPTY_PASSWORD,
+                    "Button",BS_AUTOCHECKBOX | NOT WS_VISIBLE | WS_TABSTOP,7,217,205,10
 END
 
 IDD_LANGUAGE DIALOGEX 0, 0, 209, 183
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
-CAPTION "TrueCrypt - Language"
+CAPTION "VeraCrypt - Language"
 FONT 8, "MS Shell Dlg", 400, 0, 0x1
 BEGIN
     LISTBOX         IDC_LANGLIST,6,7,197,67,LBS_SORT | LBS_NOINTEGRALHEIGHT | WS_VSCROLL | WS_TABSTOP
@@ -130,7 +140,7 @@
 
 IDD_BENCHMARK_DLG DIALOGEX 0, 0, 330, 223
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
-CAPTION "TrueCrypt - Encryption Algorithm Benchmark"
+CAPTION "VeraCrypt - Encryption Algorithm Benchmark"
 FONT 8, "MS Shell Dlg", 400, 0, 0x1
 BEGIN
     COMBOBOX        IDC_BENCHMARK_BUFFER_SIZE,55,7,77,129,CBS_DROPDOWNLIST | CBS_AUTOHSCROLL | WS_VSCROLL | WS_TABSTOP
@@ -151,7 +161,7 @@
 
 IDD_CIPHER_TEST_DLG DIALOGEX 0, 0, 326, 249
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
-CAPTION "TrueCrypt - Test Vectors"
+CAPTION "VeraCrypt - Test Vectors"
 FONT 8, "MS Shell Dlg", 0, 0, 0x0
 BEGIN
     COMBOBOX        IDC_CIPHER,109,10,104,126,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
@@ -172,9 +182,9 @@
     GROUPBOX        "Plaintext (hexadecimal)",IDT_TEST_PLAINTEXT,1,140,323,33
     GROUPBOX        "Ciphertext (hexadecimal)",IDT_TEST_CIPHERTEXT,1,174,323,33
     RTEXT           "",IDC_TESTS_MESSAGE,50,213,178,10
-    CONTROL         "",IDC_REDTICK,"REDTICK",0x0,234,214,10,8
+    CONTROL         "",IDC_REDTICK,"VCREDTICK",0x0,234,214,10,8
     RTEXT           "Key size:",IDT_KEY,8,57,56,8
-    RTEXT           "Plaintext size:",IDT_PLAINTEXT,190,153,63,8
+    RTEXT           "Plaintext size:",IDT_PLAINTEXT,174,153,79,8
     LTEXT           "bits",IDT_KEY_UNIT,114,57,45,8
     RTEXT           "Cipher:",IDT_CIPHER,38,13,68,8
     LTEXT           "bits",IDT_PLAINTEXT_SIZE_UNIT,298,153,22,8
@@ -191,27 +201,37 @@
 BEGIN
     DEFPUSHBUTTON   "OK",IDOK,305,200,58,14
     PUSHBUTTON      "&Print",IDC_PRINT,156,200,58,14
-    CONTROL         "",IDC_INFO_BOX_TEXT,"RichEdit20A",ES_MULTILINE | ES_READONLY | ES_NUMBER | WS_BORDER | WS_VSCROLL | WS_TABSTOP,5,6,361,188
+    CONTROL         "",IDC_INFO_BOX_TEXT,"RichEdit20W",ES_MULTILINE | ES_READONLY | ES_NUMBER | WS_BORDER | WS_VSCROLL | WS_TABSTOP,5,6,361,188
 END
 
-IDD_KEYFILE_GENERATOR DIALOGEX 0, 0, 308, 270
+IDD_KEYFILE_GENERATOR DIALOGEX 0, 0, 357, 362
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
-CAPTION "TrueCrypt - Keyfile Generator"
+CAPTION "VeraCrypt - Keyfile Generator"
 FONT 8, "MS Shell Dlg", 400, 0, 0x1
 BEGIN
-    DEFPUSHBUTTON   "Close",IDCLOSE,237,10,59,14
-    COMBOBOX        IDC_PRF_ID,79,49,91,90,CBS_DROPDOWNLIST | WS_TABSTOP
-    PUSHBUTTON      "Generate and Save Keyfile...",IDC_GENERATE_AND_SAVE_KEYFILE,89,248,131,14
-    LTEXT           "IMPORTANT: Move your mouse as randomly as possible within this window. The longer you move it, the better. This significantly increases the cryptographic strength of the keyfile.",IDT_KEYFILE_GENERATOR_NOTE,11,5,213,33
-    CONTROL         "",IDC_STATIC,"Static",SS_ETCHEDHORZ,1,40,307,1,WS_EX_STATICEDGE
-    RTEXT           "Mixing PRF:",IDT_PRF,6,51,67,10,SS_CENTERIMAGE
-    GROUPBOX        "Current Pool Content",IDT_POOL_CONTENTS,6,70,296,170
-    CONTROL         "",IDC_POOL_CONTENTS,"Static",SS_LEFTNOWORDWRAP | WS_GROUP,16,83,282,148,WS_EX_TRANSPARENT
-    CONTROL         "Display pool content",IDC_DISPLAY_POOL_CONTENTS,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,191,51,111,10
+    DEFPUSHBUTTON   "Close",IDCLOSE,291,10,59,14
+    COMBOBOX        IDC_PRF_ID,97,49,91,90,CBS_DROPDOWNLIST | WS_TABSTOP
+    CONTROL         "Display pool content",IDC_DISPLAY_POOL_CONTENTS,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,209,51,141,10
+    EDITTEXT        IDC_NUMBER_KEYFILES,124,278,51,14,ES_RIGHT | ES_AUTOHSCROLL | ES_NUMBER
+    EDITTEXT        IDC_KEYFILES_SIZE,124,298,51,14,ES_RIGHT | ES_AUTOHSCROLL | ES_NUMBER
+    CONTROL         "Random size ( 64 <-> 1048576 )",IDC_KEYFILES_RANDOM_SIZE,
+                    "Button",BS_AUTOCHECKBOX | WS_TABSTOP,181,300,174,10
+    EDITTEXT        IDC_KEYFILES_BASE_NAME,124,318,141,14,ES_AUTOHSCROLL
+    PUSHBUTTON      "Generate and Save Keyfile...",IDC_GENERATE_AND_SAVE_KEYFILE,124,336,141,14
+    LTEXT           "IMPORTANT: Move your mouse as randomly as possible within this window. The longer you move it, the better. This significantly increases the cryptographic strength of the keyfile.",IDT_KEYFILE_GENERATOR_NOTE,11,5,271,33
+    CONTROL         "",IDC_STATIC,"Static",SS_ETCHEDHORZ,1,40,356,1,WS_EX_STATICEDGE
+    RTEXT           "Mixing PRF:",IDT_PRF,7,51,85,10,SS_CENTERIMAGE
+    GROUPBOX        "Current Pool Content",IDT_POOL_CONTENTS,6,70,344,170
+    CTEXT           "",IDC_POOL_CONTENTS,16,83,323,148,0,WS_EX_TRANSPARENT
+    RTEXT           "Number of keyfiles:",IDT_NUMBER_KEYFILES,9,281,110,8
+    RTEXT           "Keyfiles base name:",IDT_KEYFILES_BASE_NAME,9,321,110,8
+    RTEXT           "Keyfiles size (in Bytes):",IDT_KEYFILES_SIZE,9,300,110,8
+    CONTROL         "",IDC_ENTROPY_BAR,"msctls_progress32",PBS_SMOOTH | WS_BORDER,18,255,321,12
+    GROUPBOX        "Randomness Collected From Mouse Movements",IDT_ENTROPY_BAR,6,244,344,29
 END
 
 IDD_MULTI_CHOICE_DLG DIALOGEX 0, 0, 167, 322
-STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
+STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION
 FONT 8, "MS Shell Dlg", 0, 0, 0x0
 BEGIN
     PUSHBUTTON      "",IDC_CHOICE10,7,292,153,24,BS_CENTER | BS_MULTILINE,WS_EX_STATICEDGE
@@ -275,12 +295,12 @@
     GROUPBOX        "",IDC_STATIC,5,2,228,51
 END
 
-IDD_RANDOM_POOL_ENRICHMENT DIALOGEX 0, 0, 308, 270
+IDD_RANDOM_POOL_ENRICHMENT DIALOGEX 0, 0, 308, 301
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION | WS_SYSMENU
-CAPTION "TrueCrypt - Random Pool Enrichment"
+CAPTION "VeraCrypt - Random Pool Enrichment"
 FONT 8, "MS Shell Dlg", 400, 0, 0x1
 BEGIN
-    DEFPUSHBUTTON   "&Continue",IDC_CONTINUE,119,248,71,14
+    DEFPUSHBUTTON   "&Continue",IDC_CONTINUE,119,278,71,14
     COMBOBOX        IDC_PRF_ID,79,49,91,90,CBS_DROPDOWNLIST | WS_TABSTOP
     LTEXT           "IMPORTANT: Move your mouse as randomly as possible within this window. The longer you move it, the better. This significantly increases security. When done, click 'Continue'.",IDT_RANDOM_POOL_ENRICHMENT_NOTE,11,6,282,25
     CONTROL         "",IDC_STATIC,"Static",SS_ETCHEDHORZ,1,37,307,1,WS_EX_STATICEDGE
@@ -288,17 +308,28 @@
     GROUPBOX        "Current Pool Content",IDT_POOL_CONTENTS,6,70,296,170
     CONTROL         "",IDC_POOL_CONTENTS,"Static",SS_LEFTNOWORDWRAP | WS_GROUP,16,83,282,148,WS_EX_TRANSPARENT
     CONTROL         "Display pool content",IDC_DISPLAY_POOL_CONTENTS,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,191,51,111,10
+    CONTROL         "",IDC_ENTROPY_BAR,"msctls_progress32",PBS_SMOOTH | WS_BORDER,16,255,275,12
+    GROUPBOX        "Randomness Collected From Mouse Movements",IDT_ENTROPY_BAR,7,244,294,29
 END
 
 IDD_STATIC_MODELESS_WAIT_DLG DIALOGEX 0, 0, 292, 42
 STYLE DS_SYSMODAL | DS_SETFONT | DS_MODALFRAME | DS_SETFOREGROUND | DS_3DLOOK | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION
 EXSTYLE WS_EX_TOPMOST | WS_EX_TOOLWINDOW
-CAPTION "TrueCrypt"
+CAPTION "VeraCrypt"
 FONT 8, "MS Shell Dlg", 0, 0, 0x0
 BEGIN
     LTEXT           "Please wait. This process may take a long time...",IDT_STATIC_MODELESS_WAIT_DLG_INFO,9,8,274,9
 END
 
+IDD_STATIC_MODAL_WAIT_DLG DIALOGEX 0, 0, 292, 74
+STYLE DS_SETFONT | DS_MODALFRAME | DS_SETFOREGROUND | DS_FIXEDSYS | DS_CENTER | WS_POPUP | WS_CAPTION
+CAPTION "VeraCrypt"
+FONT 8, "MS Shell Dlg", 0, 0, 0x0
+BEGIN
+    CTEXT           "Please wait...\nThis process may take a long time and VeraCrypt may seem unresponsive.",IDT_STATIC_MODAL_WAIT_DLG_INFO,9,11,274,33
+    CONTROL         "",IDC_WAIT_PROGRESS_BAR,"msctls_progress32",WS_BORDER,7,49,278,14
+END
+
 
 /////////////////////////////////////////////////////////////////////////////
 //
@@ -308,9 +339,13 @@
 #ifdef APSTUDIO_INVOKED
 GUIDELINES DESIGNINFO 
 BEGIN
+    IDD_ABOUT_DLG, DIALOG
+    BEGIN
+    END
+
     IDD_COMMANDHELP_DLG, DIALOG
     BEGIN
-        BOTTOMMARGIN, 205
+        BOTTOMMARGIN, 266
     END
 
     IDD_RAWDEVICES_DLG, DIALOG
@@ -325,7 +360,7 @@
     BEGIN
         LEFTMARGIN, 7
         TOPMARGIN, 7
-        BOTTOMMARGIN, 166
+        BOTTOMMARGIN, 225
     END
 
     IDD_KEYFILES, DIALOG
@@ -333,7 +368,7 @@
         LEFTMARGIN, 7
         RIGHTMARGIN, 330
         TOPMARGIN, 7
-        BOTTOMMARGIN, 230
+        BOTTOMMARGIN, 244
     END
 
     IDD_LANGUAGE, DIALOG
@@ -371,9 +406,8 @@
     IDD_KEYFILE_GENERATOR, DIALOG
     BEGIN
         LEFTMARGIN, 7
-        RIGHTMARGIN, 299
         TOPMARGIN, 7
-        BOTTOMMARGIN, 266
+        BOTTOMMARGIN, 358
     END
 
     IDD_MULTI_CHOICE_DLG, DIALOG
@@ -421,7 +455,7 @@
         LEFTMARGIN, 7
         RIGHTMARGIN, 301
         TOPMARGIN, 7
-        BOTTOMMARGIN, 267
+        BOTTOMMARGIN, 298
     END
 
     IDD_STATIC_MODELESS_WAIT_DLG, DIALOG
@@ -431,6 +465,14 @@
         TOPMARGIN, 7
         BOTTOMMARGIN, 35
     END
+
+    IDD_STATIC_MODAL_WAIT_DLG, DIALOG
+    BEGIN
+        LEFTMARGIN, 7
+        RIGHTMARGIN, 285
+        TOPMARGIN, 7
+        BOTTOMMARGIN, 67
+    END
 END
 #endif    // APSTUDIO_INVOKED
 
@@ -449,6 +491,17 @@
 IDR_BOOT_LOADER_AES     BIN                     "..\\Boot\\Windows\\Release_AES\\BootLoader.com.gz"
 IDR_BOOT_LOADER_SERPENT BIN                     "..\\Boot\\Windows\\Release_Serpent\\BootLoader.com.gz"
 IDR_BOOT_LOADER_TWOFISH BIN                     "..\\Boot\\Windows\\Release_Twofish\\BootLoader.com.gz"
+
+IDR_BOOT_SECTOR_SHA2    BIN                     "..\\Boot\\Windows\\Release_SHA2\\BootSector.bin"
+IDR_BOOT_SECTOR_AES_SHA2 BIN                     "..\\Boot\\Windows\\Release_AES_SHA2\\BootSector.bin"
+IDR_BOOT_SECTOR_SERPENT_SHA2 BIN                     "..\\Boot\\Windows\\Release_Serpent_SHA2\\BootSector.bin"
+IDR_BOOT_SECTOR_TWOFISH_SHA2 BIN                     "..\\Boot\\Windows\\Release_Twofish_SHA2\\BootSector.bin"
+IDR_BOOT_LOADER_SHA2    BIN                     "..\\Boot\\Windows\\Release_SHA2\\BootLoader.com.gz"
+IDR_BOOT_LOADER_AES_SHA2 BIN                     "..\\Boot\\Windows\\Release_AES_SHA2\\BootLoader.com.gz"
+IDR_BOOT_LOADER_SERPENT_SHA2 BIN                     "..\\Boot\\Windows\\Release_Serpent_SHA2\\BootLoader.com.gz"
+IDR_BOOT_LOADER_TWOFISH_SHA2 BIN                     "..\\Boot\\Windows\\Release_Twofish_SHA2\\BootLoader.com.gz"
+
+
 IDR_RESCUE_BOOT_SECTOR  BIN                     "..\\Boot\\Windows\\Rescue\\BootSector.bin"
 IDR_RESCUE_BOOT_SECTOR_AES BIN                     "..\\Boot\\Windows\\Rescue_AES\\BootSector.bin"
 IDR_RESCUE_BOOT_SECTOR_SERPENT BIN                     "..\\Boot\\Windows\\Rescue_Serpent\\BootSector.bin"
@@ -458,6 +511,15 @@
 IDR_RESCUE_LOADER_SERPENT BIN                     "..\\Boot\\Windows\\Rescue_Serpent\\BootLoader.com.gz"
 IDR_RESCUE_LOADER_TWOFISH BIN                     "..\\Boot\\Windows\\Rescue_Twofish\\BootLoader.com.gz"
 
+IDR_RESCUE_BOOT_SECTOR_SHA2 BIN                     "..\\Boot\\Windows\\Rescue_SHA2\\BootSector.bin"
+IDR_RESCUE_BOOT_SECTOR_AES_SHA2 BIN                     "..\\Boot\\Windows\\Rescue_AES_SHA2\\BootSector.bin"
+IDR_RESCUE_BOOT_SECTOR_SERPENT_SHA2 BIN                     "..\\Boot\\Windows\\Rescue_Serpent_SHA2\\BootSector.bin"
+IDR_RESCUE_BOOT_SECTOR_TWOFISH_SHA2 BIN                     "..\\Boot\\Windows\\Rescue_Twofish_SHA2\\BootSector.bin"
+IDR_RESCUE_LOADER_SHA2  BIN                     "..\\Boot\\Windows\\Rescue_SHA2\\BootLoader.com.gz"
+IDR_RESCUE_LOADER_AES_SHA2 BIN                     "..\\Boot\\Windows\\Rescue_AES_SHA2\\BootLoader.com.gz"
+IDR_RESCUE_LOADER_SERPENT_SHA2 BIN                     "..\\Boot\\Windows\\Rescue_Serpent_SHA2\\BootLoader.com.gz"
+IDR_RESCUE_LOADER_TWOFISH_SHA2 BIN                     "..\\Boot\\Windows\\Rescue_Twofish_SHA2\\BootLoader.com.gz"
+
 /////////////////////////////////////////////////////////////////////////////
 //
 // XML
@@ -512,9 +574,9 @@
 
 // Icon with lowest ID value placed first to ensure application icon
 // remains consistent on all systems.
-IDI_TRUECRYPT_ICON      ICON                    "..\\Common\\TrueCrypt.ico"
-IDI_TRUECRYPT_VOL_ICON  ICON                    "..\\Common\\TrueCrypt_volume.ico"
-IDI_TRUECRYPT_MOUNTED_ICON ICON                    "..\\Common\\TrueCrypt_mounted.ico"
+IDI_TRUECRYPT_ICON      ICON                    "..\\Common\\VeraCrypt.ico"
+IDI_TRUECRYPT_VOL_ICON  ICON                    "..\\Common\\VeraCrypt_volume.ico"
+IDI_TRUECRYPT_MOUNTED_ICON ICON                    "..\\Common\\VeraCrypt_mounted.ico"
 
 /////////////////////////////////////////////////////////////////////////////
 //
