--- ./Main/Resources.cpp	2012-02-07 14:36:48.000000000 +0400
+++ ../veracrypt_1.17_Source/src/Main/Resources.cpp	2016-02-14 12:11:21.000000000 +0300
@@ -1,9 +1,13 @@
 /*
- Copyright (c) 2008 TrueCrypt Developers Association. All rights reserved.
-
- Governed by the TrueCrypt License 3.0 the full text of which is contained in
- the file License.txt included in TrueCrypt binary and source code distribution
- packages.
+ Derived from source code of TrueCrypt 7.1a, which is
+ Copyright (c) 2008-2012 TrueCrypt Developers Association and which is governed
+ by the TrueCrypt License 3.0.
+
+ Modifications and additions to the original source code (contained in this file) 
+ and all other portions of this file are Copyright (c) 2013-2016 IDRIX
+ and are governed by the Apache License 2.0 the full text of which is
+ contained in the file License.txt included in VeraCrypt binary and source
+ code distribution packages.
 */
 
 #include "System.h"
@@ -14,7 +18,7 @@
 #include "Main/resource.h"
 #endif
 
-namespace TrueCrypt
+namespace VeraCrypt
 {
 
 #ifdef TC_WINDOWS
@@ -146,13 +150,13 @@
 #endif
 	}
 
-	wxIcon Resources::GetTrueCryptIcon ()
+	wxIcon Resources::GetVeraCryptIcon ()
 	{
 #ifdef TC_WINDOWS
 		return wxIcon (L"IDI_TRUECRYPT_ICON", wxBITMAP_TYPE_ICO_RESOURCE, 16, 16);
 #else
-#		include "Resources/Icons/TrueCrypt-16x16.xpm"
-		return wxIcon (TrueCryptIcon16x16);
+#		include "Resources/Icons/VeraCrypt-256x256.xpm"
+		return wxIcon (VeraCryptIcon256x256);
 #endif
 	}
 
@@ -163,7 +167,7 @@
 #else
 		static const byte VolumeWizardIcon[] =
 		{
-#			include "Format/TrueCrypt_Wizard.bmp.h"
+#			include "Format/VeraCrypt_Wizard.bmp.h"
 		};
 
 		wxMemoryInputStream stream (VolumeWizardIcon, sizeof (VolumeWizardIcon));
