--- ./Format/Format.vcproj	2012-02-07 14:37:04.000000000 +0400
+++ ../veracrypt_1.17_Source/src/Format/Format.vcproj	2016-02-14 12:11:21.000000000 +0300
@@ -12,6 +12,9 @@
 		<Platform
 			Name="Win32"
 		/>
+		<Platform
+			Name="x64"
+		/>
 	</Platforms>
 	<ToolFiles>
 	</ToolFiles>
@@ -22,7 +25,7 @@
 			IntermediateDirectory="Debug"
 			ConfigurationType="1"
 			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
-			CharacterSet="2"
+			CharacterSet="1"
 			>
 			<Tool
 				Name="VCPreBuildEventTool"
@@ -43,7 +46,7 @@
 			<Tool
 				Name="VCCLCompilerTool"
 				Optimization="0"
-				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;$(PKCS11_INC)"
+				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;..\PKCS11"
 				PreprocessorDefinitions="VOLFORMAT;WIN32;DEBUG;_DEBUG;_WINDOWS;_CRT_SECURE_NO_DEPRECATE;_CRT_NON_CONFORMING_SWPRINTFS"
 				MinimalRebuild="true"
 				BasicRuntimeChecks="3"
@@ -65,11 +68,12 @@
 			/>
 			<Tool
 				Name="VCLinkerTool"
-				AdditionalDependencies="comctl32.lib setupapi.lib ..\Crypto\Debug\crypto.lib"
-				OutputFile="$(OutDir)/TrueCryptFormat.exe"
+				AdditionalDependencies="..\Crypto\Debug\crypto.lib mpr.lib"
+				OutputFile="$(OutDir)/VeraCryptFormat.exe"
 				LinkIncremental="2"
 				GenerateManifest="false"
 				IgnoreAllDefaultLibraries="false"
+				DelayLoadDLLs="mpr.dll"
 				GenerateDebugInformation="true"
 				ProgramDatabaseFile="$(OutDir)/Format.pdb"
 				SubSystem="2"
@@ -98,7 +102,94 @@
 			/>
 			<Tool
 				Name="VCPostBuildEventTool"
-				CommandLine="md &quot;..\Debug\Setup Files&quot; 2&gt;NUL:&#x0D;&#x0A;copy Debug\TrueCryptFormat.exe &quot;..\Debug\Setup Files\TrueCrypt Format.exe&quot; &gt;NUL:&#x0D;&#x0A;"
+				CommandLine="md &quot;..\Debug\Setup Files&quot; 2&gt;NUL:&#x0D;&#x0A;copy Debug\VeraCryptFormat.exe &quot;..\Debug\Setup Files\VeraCrypt Format.exe&quot; &gt;NUL:&#x0D;&#x0A;"
+			/>
+		</Configuration>
+		<Configuration
+			Name="Debug|x64"
+			OutputDirectory="$(PlatformName)\$(ConfigurationName)"
+			IntermediateDirectory="$(PlatformName)\$(ConfigurationName)"
+			ConfigurationType="1"
+			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
+			CharacterSet="1"
+			>
+			<Tool
+				Name="VCPreBuildEventTool"
+			/>
+			<Tool
+				Name="VCCustomBuildTool"
+			/>
+			<Tool
+				Name="VCXMLDataGeneratorTool"
+			/>
+			<Tool
+				Name="VCWebServiceProxyGeneratorTool"
+			/>
+			<Tool
+				Name="VCMIDLTool"
+				TargetEnvironment="3"
+				TypeLibraryName="$(SolutionDir)/$(ProjectName)/$(ProjectName).tlb"
+			/>
+			<Tool
+				Name="VCCLCompilerTool"
+				Optimization="0"
+				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;..\PKCS11"
+				PreprocessorDefinitions="VOLFORMAT;WIN32;DEBUG;_DEBUG;_WINDOWS;_CRT_SECURE_NO_DEPRECATE;_CRT_NON_CONFORMING_SWPRINTFS"
+				MinimalRebuild="true"
+				BasicRuntimeChecks="3"
+				RuntimeLibrary="1"
+				BufferSecurityCheck="true"
+				UsePrecompiledHeader="0"
+				WarningLevel="4"
+				DebugInformationFormat="3"
+				DisableSpecificWarnings="4057;4100;4127;4201;4204;4701;4706"
+			/>
+			<Tool
+				Name="VCManagedResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCPreLinkEventTool"
+			/>
+			<Tool
+				Name="VCLinkerTool"
+				AdditionalDependencies="..\Crypto\x64\Debug\crypto.lib mpr.lib"
+				OutputFile="$(OutDir)/VeraCryptFormat.exe"
+				LinkIncremental="2"
+				GenerateManifest="false"
+				IgnoreAllDefaultLibraries="false"
+				DelayLoadDLLs="mpr.dll"
+				GenerateDebugInformation="true"
+				ProgramDatabaseFile="$(OutDir)/Format.pdb"
+				SubSystem="2"
+				RandomizedBaseAddress="1"
+				DataExecutionPrevention="2"
+				TargetMachine="17"
+			/>
+			<Tool
+				Name="VCALinkTool"
+			/>
+			<Tool
+				Name="VCManifestTool"
+				AdditionalManifestFiles="Format.manifest"
+			/>
+			<Tool
+				Name="VCXDCMakeTool"
+			/>
+			<Tool
+				Name="VCBscMakeTool"
+			/>
+			<Tool
+				Name="VCFxCopTool"
+			/>
+			<Tool
+				Name="VCAppVerifierTool"
+			/>
+			<Tool
+				Name="VCPostBuildEventTool"
+				CommandLine="md &quot;..\Debug\Setup Files&quot; 2&gt;NUL:&#x0D;&#x0A;copy $(TargetPath) &quot;..\Debug\Setup Files\VeraCrypt Format-x64.exe&quot; &gt;NUL:&#x0D;&#x0A;"
 			/>
 		</Configuration>
 		<Configuration
@@ -107,7 +198,7 @@
 			IntermediateDirectory="Release"
 			ConfigurationType="1"
 			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
-			CharacterSet="2"
+			CharacterSet="1"
 			>
 			<Tool
 				Name="VCPreBuildEventTool"
@@ -129,7 +220,7 @@
 				Name="VCCLCompilerTool"
 				AdditionalOptions="/w34189"
 				Optimization="2"
-				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;$(PKCS11_INC)"
+				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;..\PKCS11"
 				PreprocessorDefinitions="VOLFORMAT;WIN32;NDEBUG;_WINDOWS;_CRT_SECURE_NO_DEPRECATE;_CRT_NON_CONFORMING_SWPRINTFS"
 				RuntimeLibrary="0"
 				BufferSecurityCheck="true"
@@ -151,11 +242,12 @@
 			/>
 			<Tool
 				Name="VCLinkerTool"
-				AdditionalDependencies="comctl32.lib setupapi.lib ..\Crypto\Release\crypto.lib"
-				OutputFile="$(OutDir)/TrueCryptFormat.exe"
+				AdditionalDependencies="..\Crypto\Release\crypto.lib mpr.lib"
+				OutputFile="$(OutDir)/VeraCryptFormat.exe"
 				LinkIncremental="1"
 				GenerateManifest="false"
 				IgnoreAllDefaultLibraries="false"
+				DelayLoadDLLs="mpr.dll"
 				GenerateDebugInformation="false"
 				GenerateMapFile="true"
 				SubSystem="2"
@@ -186,7 +278,97 @@
 			/>
 			<Tool
 				Name="VCPostBuildEventTool"
-				CommandLine="copy Release\TrueCryptFormat.exe &quot;..\Release\Setup Files\TrueCrypt Format.exe&quot;"
+				CommandLine="copy Release\VeraCryptFormat.exe &quot;..\Release\Setup Files\VeraCrypt Format.exe&quot;"
+			/>
+		</Configuration>
+		<Configuration
+			Name="Release|x64"
+			OutputDirectory="$(PlatformName)\$(ConfigurationName)"
+			IntermediateDirectory="$(PlatformName)\$(ConfigurationName)"
+			ConfigurationType="1"
+			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
+			CharacterSet="1"
+			>
+			<Tool
+				Name="VCPreBuildEventTool"
+			/>
+			<Tool
+				Name="VCCustomBuildTool"
+			/>
+			<Tool
+				Name="VCXMLDataGeneratorTool"
+			/>
+			<Tool
+				Name="VCWebServiceProxyGeneratorTool"
+			/>
+			<Tool
+				Name="VCMIDLTool"
+				TargetEnvironment="3"
+				TypeLibraryName="$(SolutionDir)/$(ProjectName)/$(ProjectName).tlb"
+			/>
+			<Tool
+				Name="VCCLCompilerTool"
+				AdditionalOptions="/w34189"
+				Optimization="2"
+				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;..\PKCS11"
+				PreprocessorDefinitions="VOLFORMAT;WIN32;NDEBUG;_WINDOWS;_CRT_SECURE_NO_DEPRECATE;_CRT_NON_CONFORMING_SWPRINTFS"
+				RuntimeLibrary="0"
+				BufferSecurityCheck="true"
+				UsePrecompiledHeader="0"
+				AssemblerOutput="2"
+				AssemblerListingLocation="$(IntDir)/"
+				WarningLevel="4"
+				DebugInformationFormat="0"
+				DisableSpecificWarnings="4057;4100;4127;4201;4204;4701;4706"
+			/>
+			<Tool
+				Name="VCManagedResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCPreLinkEventTool"
+			/>
+			<Tool
+				Name="VCLinkerTool"
+				AdditionalDependencies="..\Crypto\x64\Release\crypto.lib mpr.lib"
+				OutputFile="$(OutDir)/VeraCryptFormat.exe"
+				LinkIncremental="1"
+				GenerateManifest="false"
+				IgnoreAllDefaultLibraries="false"
+				DelayLoadDLLs="mpr.dll"
+				GenerateDebugInformation="false"
+				GenerateMapFile="true"
+				SubSystem="2"
+				OptimizeReferences="2"
+				EnableCOMDATFolding="2"
+				RandomizedBaseAddress="1"
+				DataExecutionPrevention="2"
+				TargetMachine="17"
+			/>
+			<Tool
+				Name="VCALinkTool"
+			/>
+			<Tool
+				Name="VCManifestTool"
+				AdditionalManifestFiles="Format.manifest"
+			/>
+			<Tool
+				Name="VCXDCMakeTool"
+			/>
+			<Tool
+				Name="VCBscMakeTool"
+			/>
+			<Tool
+				Name="VCFxCopTool"
+			/>
+			<Tool
+				Name="VCAppVerifierTool"
+			/>
+			<Tool
+				Name="VCPostBuildEventTool"
+				CommandLine="copy $(TargetPath) &quot;..\Release\Setup Files\VeraCrypt Format-x64.exe&quot;"
 			/>
 		</Configuration>
 	</Configurations>
@@ -218,6 +400,14 @@
 					/>
 				</FileConfiguration>
 				<FileConfiguration
+					Name="Debug|x64"
+					>
+					<Tool
+						Name="VCCLCompilerTool"
+						CompileAs="2"
+					/>
+				</FileConfiguration>
+				<FileConfiguration
 					Name="Release|Win32"
 					>
 					<Tool
@@ -225,6 +415,14 @@
 						CompileAs="2"
 					/>
 				</FileConfiguration>
+				<FileConfiguration
+					Name="Release|x64"
+					>
+					<Tool
+						Name="VCCLCompilerTool"
+						CompileAs="2"
+					/>
+				</FileConfiguration>
 			</File>
 			<File
 				RelativePath=".\Tcformat.c"
@@ -238,6 +436,14 @@
 					/>
 				</FileConfiguration>
 				<FileConfiguration
+					Name="Debug|x64"
+					>
+					<Tool
+						Name="VCCLCompilerTool"
+						CompileAs="2"
+					/>
+				</FileConfiguration>
+				<FileConfiguration
 					Name="Release|Win32"
 					>
 					<Tool
@@ -245,6 +451,14 @@
 						CompileAs="2"
 					/>
 				</FileConfiguration>
+				<FileConfiguration
+					Name="Release|x64"
+					>
+					<Tool
+						Name="VCCLCompilerTool"
+						CompileAs="2"
+					/>
+				</FileConfiguration>
 			</File>
 			<Filter
 				Name="Common"
@@ -285,6 +499,14 @@
 						/>
 					</FileConfiguration>
 					<FileConfiguration
+						Name="Debug|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
+					<FileConfiguration
 						Name="Release|Win32"
 						>
 						<Tool
@@ -292,6 +514,14 @@
 							CompileAs="2"
 						/>
 					</FileConfiguration>
+					<FileConfiguration
+						Name="Release|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
 				</File>
 				<File
 					RelativePath="..\Common\Dlgcode.c"
@@ -305,6 +535,14 @@
 						/>
 					</FileConfiguration>
 					<FileConfiguration
+						Name="Debug|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
+					<FileConfiguration
 						Name="Release|Win32"
 						>
 						<Tool
@@ -312,6 +550,14 @@
 							CompileAs="2"
 						/>
 					</FileConfiguration>
+					<FileConfiguration
+						Name="Release|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
 				</File>
 				<File
 					RelativePath="..\Common\EncryptionThreadPool.c"
@@ -345,6 +591,14 @@
 						/>
 					</FileConfiguration>
 					<FileConfiguration
+						Name="Debug|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
+					<FileConfiguration
 						Name="Release|Win32"
 						>
 						<Tool
@@ -352,6 +606,14 @@
 							CompileAs="2"
 						/>
 					</FileConfiguration>
+					<FileConfiguration
+						Name="Release|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
 				</File>
 				<File
 					RelativePath="..\Common\Language.c"
@@ -539,6 +801,38 @@
 			UniqueIdentifier="{67DA6AB6-F800-4c08-8B7A-83BB121AAD01}"
 			>
 			<File
+				RelativePath="..\Boot\Windows\Release_Serpent_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Rescue_Twofish_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Rescue_Serpent_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Rescue_AES_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Rescue_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Release_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Release_AES_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Release_Twofish_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
 				RelativePath=".\Format.manifest"
 				>
 			</File>
@@ -552,46 +846,50 @@
 				>
 			</File>
 			<File
-				RelativePath="..\Common\TrueCrypt_Volume.ico"
+				RelativePath="..\Common\VeraCrypt_mounted.ico"
 				>
 			</File>
 			<File
-				RelativePath=".\TrueCrypt_wizard.bmp"
+				RelativePath="..\Common\VeraCrypt_Volume.ico"
+				>
+			</File>
+			<File
+				RelativePath=".\VeraCrypt_wizard.bmp"
 				>
 			</File>
 			<Filter
 				Name="Common"
 				>
 				<File
-					RelativePath="..\Boot\Windows\Rescue_AES\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Release_AES\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Rescue\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Release\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Rescue_Twofish\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Release_Twofish\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Release_Serpent\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Rescue_Serpent\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Release_AES\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Rescue_AES\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Release\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Rescue\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Release_Twofish\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Rescue_Twofish\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Rescue_Serpent\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Release_Serpent\BootLoader.com.gz"
 					>
 				</File>
 				<File
@@ -638,6 +936,14 @@
 						/>
 					</FileConfiguration>
 					<FileConfiguration
+						Name="Debug|x64"
+						ExcludedFromBuild="true"
+						>
+						<Tool
+							Name="VCResourceCompilerTool"
+						/>
+					</FileConfiguration>
+					<FileConfiguration
 						Name="Release|Win32"
 						ExcludedFromBuild="true"
 						>
@@ -645,6 +951,14 @@
 							Name="VCResourceCompilerTool"
 						/>
 					</FileConfiguration>
+					<FileConfiguration
+						Name="Release|x64"
+						ExcludedFromBuild="true"
+						>
+						<Tool
+							Name="VCResourceCompilerTool"
+						/>
+					</FileConfiguration>
 				</File>
 				<File
 					RelativePath="..\Boot\Windows\Release\Decompressor.com"
@@ -671,11 +985,43 @@
 					>
 				</File>
 				<File
-					RelativePath="..\Common\TrueCrypt.ico"
+					RelativePath="..\Common\VeraCrypt.ico"
 					>
 				</File>
 			</Filter>
 		</Filter>
+		<File
+			RelativePath="..\Boot\Windows\Release_Serpent_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Release_Twofish_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Rescue_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Rescue_AES_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Rescue_Serpent_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Rescue_Twofish_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Release_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Release_AES_SHA2\BootLoader.com.gz"
+			>
+		</File>
 	</Files>
 	<Globals>
 	</Globals>
