--- ./Driver/Driver.rc	2012-02-07 14:36:48.000000000 +0400
+++ ../veracrypt_1.17_Source/src/Driver/Driver.rc	2016-02-14 12:11:21.000000000 +0300
@@ -27,8 +27,8 @@
 //
 
 VS_VERSION_INFO VERSIONINFO
- FILEVERSION 7,1,1,0
- PRODUCTVERSION 7,1,1,0
+ FILEVERSION 1,17,22,0
+ PRODUCTVERSION 1,17,22,0
  FILEFLAGSMASK 0x17L
 #ifdef _DEBUG
  FILEFLAGS 0x1L
@@ -43,13 +43,13 @@
     BEGIN
         BLOCK "040904b0"
         BEGIN
-            VALUE "CompanyName", "TrueCrypt Foundation"
-            VALUE "FileDescription", "TrueCrypt Driver"
-            VALUE "FileVersion", "7.1a"
-            VALUE "LegalTrademarks", "TrueCrypt"
-            VALUE "OriginalFilename", "truecrypt.sys"
-            VALUE "ProductName", "TrueCrypt"
-            VALUE "ProductVersion", "7.1a"
+            VALUE "CompanyName", "IDRIX"
+            VALUE "FileDescription", "VeraCrypt Driver"
+            VALUE "FileVersion", "1.17"
+            VALUE "LegalTrademarks", "VeraCrypt"
+            VALUE "OriginalFilename", "veracrypt.sys"
+            VALUE "ProductName", "VeraCrypt"
+            VALUE "ProductVersion", "1.17"
         END
     END
     BLOCK "VarFileInfo"
