--- ./Mount/Mount.vcproj	2012-02-07 14:37:04.000000000 +0400
+++ ../veracrypt_1.17_Source/src/Mount/Mount.vcproj	2016-02-14 12:11:21.000000000 +0300
@@ -12,6 +12,9 @@
 		<Platform
 			Name="Win32"
 		/>
+		<Platform
+			Name="x64"
+		/>
 	</Platforms>
 	<ToolFiles>
 	</ToolFiles>
@@ -22,7 +25,7 @@
 			IntermediateDirectory="Debug"
 			ConfigurationType="1"
 			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
-			CharacterSet="2"
+			CharacterSet="1"
 			>
 			<Tool
 				Name="VCPreBuildEventTool"
@@ -45,7 +48,7 @@
 			<Tool
 				Name="VCCLCompilerTool"
 				Optimization="0"
-				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;$(PKCS11_INC)"
+				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;..\PKCS11"
 				PreprocessorDefinitions="TCMOUNT;WIN32;DEBUG;_DEBUG;_WINDOWS;_CRT_SECURE_NO_DEPRECATE;_CRT_NON_CONFORMING_SWPRINTFS"
 				MinimalRebuild="true"
 				ExceptionHandling="1"
@@ -71,11 +74,12 @@
 			/>
 			<Tool
 				Name="VCLinkerTool"
-				AdditionalDependencies="comctl32.lib setupapi.lib version.lib ..\Crypto\Debug\crypto.lib"
-				OutputFile="$(OutDir)/TrueCrypt.exe"
+				AdditionalDependencies="..\Crypto\Debug\crypto.lib mpr.lib"
+				OutputFile="$(OutDir)/VeraCrypt.exe"
 				LinkIncremental="2"
 				GenerateManifest="false"
 				IgnoreAllDefaultLibraries="false"
+				DelayLoadDLLs="mpr.dll"
 				GenerateDebugInformation="true"
 				ProgramDatabaseFile="$(OutDir)/Mount.pdb"
 				SubSystem="2"
@@ -104,7 +108,100 @@
 			/>
 			<Tool
 				Name="VCPostBuildEventTool"
-				CommandLine="md &quot;..\Debug\Setup Files&quot; 2&gt;NUL:&#x0D;&#x0A;copy Debug\TrueCrypt.exe &quot;..\Debug\Setup Files&quot; &gt;NUL:&#x0D;&#x0A;"
+				CommandLine="md &quot;..\Debug\Setup Files&quot; 2&gt;NUL:&#x0D;&#x0A;copy Debug\VeraCrypt.exe &quot;..\Debug\Setup Files&quot; &gt;NUL:&#x0D;&#x0A;"
+			/>
+		</Configuration>
+		<Configuration
+			Name="Debug|x64"
+			OutputDirectory="$(PlatformName)\$(ConfigurationName)"
+			IntermediateDirectory="$(PlatformName)\$(ConfigurationName)"
+			ConfigurationType="1"
+			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
+			CharacterSet="1"
+			>
+			<Tool
+				Name="VCPreBuildEventTool"
+			/>
+			<Tool
+				Name="VCCustomBuildTool"
+			/>
+			<Tool
+				Name="VCXMLDataGeneratorTool"
+			/>
+			<Tool
+				Name="VCWebServiceProxyGeneratorTool"
+			/>
+			<Tool
+				Name="VCMIDLTool"
+				AdditionalIncludeDirectories=""
+				TargetEnvironment="3"
+				TypeLibraryName="$(SolutionDir)/$(ProjectName)/$(ProjectName).tlb"
+				OutputDirectory=""
+			/>
+			<Tool
+				Name="VCCLCompilerTool"
+				Optimization="0"
+				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;..\PKCS11"
+				PreprocessorDefinitions="TCMOUNT;WIN32;DEBUG;_DEBUG;_WINDOWS;_CRT_SECURE_NO_DEPRECATE;_CRT_NON_CONFORMING_SWPRINTFS"
+				MinimalRebuild="true"
+				ExceptionHandling="1"
+				BasicRuntimeChecks="3"
+				RuntimeLibrary="1"
+				BufferSecurityCheck="true"
+				EnableFunctionLevelLinking="false"
+				UsePrecompiledHeader="0"
+				BrowseInformation="0"
+				BrowseInformationFile=""
+				WarningLevel="4"
+				DebugInformationFormat="3"
+				DisableSpecificWarnings="4057;4100;4127;4201;4701;4706"
+			/>
+			<Tool
+				Name="VCManagedResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCPreLinkEventTool"
+			/>
+			<Tool
+				Name="VCLinkerTool"
+				AdditionalDependencies="..\Crypto\x64\Debug\crypto.lib mpr.lib"
+				OutputFile="$(OutDir)/VeraCrypt.exe"
+				LinkIncremental="2"
+				GenerateManifest="false"
+				IgnoreAllDefaultLibraries="false"
+				DelayLoadDLLs="mpr.dll"
+				GenerateDebugInformation="true"
+				ProgramDatabaseFile="$(OutDir)/Mount.pdb"
+				SubSystem="2"
+				RandomizedBaseAddress="1"
+				DataExecutionPrevention="2"
+				TargetMachine="17"
+			/>
+			<Tool
+				Name="VCALinkTool"
+			/>
+			<Tool
+				Name="VCManifestTool"
+				AdditionalManifestFiles="Mount.manifest"
+			/>
+			<Tool
+				Name="VCXDCMakeTool"
+			/>
+			<Tool
+				Name="VCBscMakeTool"
+			/>
+			<Tool
+				Name="VCFxCopTool"
+			/>
+			<Tool
+				Name="VCAppVerifierTool"
+			/>
+			<Tool
+				Name="VCPostBuildEventTool"
+				CommandLine="md &quot;..\Debug\Setup Files&quot; 2&gt;NUL:&#x0D;&#x0A;copy $(TargetPath) &quot;..\Debug\Setup Files\VeraCrypt-x64.exe&quot; &gt;NUL:&#x0D;&#x0A;"
 			/>
 		</Configuration>
 		<Configuration
@@ -113,7 +210,7 @@
 			IntermediateDirectory="Release"
 			ConfigurationType="1"
 			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
-			CharacterSet="2"
+			CharacterSet="1"
 			>
 			<Tool
 				Name="VCPreBuildEventTool"
@@ -137,7 +234,7 @@
 				Name="VCCLCompilerTool"
 				AdditionalOptions="/w34189"
 				Optimization="2"
-				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;$(PKCS11_INC)"
+				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;..\PKCS11"
 				PreprocessorDefinitions="TCMOUNT;WIN32;NDEBUG;_WINDOWS;_CRT_SECURE_NO_DEPRECATE;_CRT_NON_CONFORMING_SWPRINTFS"
 				RuntimeLibrary="0"
 				BufferSecurityCheck="true"
@@ -159,11 +256,12 @@
 			/>
 			<Tool
 				Name="VCLinkerTool"
-				AdditionalDependencies="comctl32.lib setupapi.lib version.lib ..\Crypto\Release\crypto.lib"
-				OutputFile="$(OutDir)/TrueCrypt.exe"
+				AdditionalDependencies="..\Crypto\Release\crypto.lib mpr.lib"
+				OutputFile="$(OutDir)/VeraCrypt.exe"
 				LinkIncremental="1"
 				GenerateManifest="false"
 				IgnoreAllDefaultLibraries="false"
+				DelayLoadDLLs="mpr.dll"
 				GenerateDebugInformation="false"
 				GenerateMapFile="true"
 				SubSystem="2"
@@ -194,7 +292,99 @@
 			/>
 			<Tool
 				Name="VCPostBuildEventTool"
-				CommandLine="copy Release\TrueCrypt.exe &quot;..\Release\Setup Files&quot;"
+				CommandLine="copy Release\VeraCrypt.exe &quot;..\Release\Setup Files&quot;"
+			/>
+		</Configuration>
+		<Configuration
+			Name="Release|x64"
+			OutputDirectory="$(PlatformName)\$(ConfigurationName)"
+			IntermediateDirectory="$(PlatformName)\$(ConfigurationName)"
+			ConfigurationType="1"
+			InheritedPropertySheets="$(VCInstallDir)VCProjectDefaults\UpgradeFromVC71.vsprops"
+			CharacterSet="1"
+			>
+			<Tool
+				Name="VCPreBuildEventTool"
+			/>
+			<Tool
+				Name="VCCustomBuildTool"
+			/>
+			<Tool
+				Name="VCXMLDataGeneratorTool"
+			/>
+			<Tool
+				Name="VCWebServiceProxyGeneratorTool"
+			/>
+			<Tool
+				Name="VCMIDLTool"
+				AdditionalIncludeDirectories=""
+				TargetEnvironment="3"
+				TypeLibraryName="$(SolutionDir)/$(ProjectName)/$(ProjectName).tlb"
+				OutputDirectory=""
+			/>
+			<Tool
+				Name="VCCLCompilerTool"
+				AdditionalOptions="/w34189"
+				Optimization="2"
+				AdditionalIncludeDirectories="..\Common;..\Crypto;..\;..\PKCS11"
+				PreprocessorDefinitions="TCMOUNT;WIN32;NDEBUG;_WINDOWS;_CRT_SECURE_NO_DEPRECATE;_CRT_NON_CONFORMING_SWPRINTFS"
+				RuntimeLibrary="0"
+				BufferSecurityCheck="true"
+				UsePrecompiledHeader="0"
+				AssemblerOutput="2"
+				AssemblerListingLocation="$(IntDir)/"
+				WarningLevel="4"
+				DebugInformationFormat="0"
+				DisableSpecificWarnings="4057;4100;4127;4201;4701;4706"
+			/>
+			<Tool
+				Name="VCManagedResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCResourceCompilerTool"
+			/>
+			<Tool
+				Name="VCPreLinkEventTool"
+			/>
+			<Tool
+				Name="VCLinkerTool"
+				AdditionalDependencies="..\Crypto\x64\Release\crypto.lib mpr.lib"
+				OutputFile="$(OutDir)/VeraCrypt.exe"
+				LinkIncremental="1"
+				GenerateManifest="false"
+				IgnoreAllDefaultLibraries="false"
+				DelayLoadDLLs="mpr.dll"
+				GenerateDebugInformation="false"
+				GenerateMapFile="true"
+				SubSystem="2"
+				OptimizeReferences="2"
+				EnableCOMDATFolding="2"
+				RandomizedBaseAddress="1"
+				DataExecutionPrevention="2"
+				TargetMachine="17"
+			/>
+			<Tool
+				Name="VCALinkTool"
+			/>
+			<Tool
+				Name="VCManifestTool"
+				AdditionalManifestFiles="Mount.manifest"
+			/>
+			<Tool
+				Name="VCXDCMakeTool"
+			/>
+			<Tool
+				Name="VCBscMakeTool"
+			/>
+			<Tool
+				Name="VCFxCopTool"
+			/>
+			<Tool
+				Name="VCAppVerifierTool"
+			/>
+			<Tool
+				Name="VCPostBuildEventTool"
+				CommandLine="copy $(TargetPath) &quot;..\Release\Setup Files\VeraCrypt-x64.exe&quot;"
 			/>
 		</Configuration>
 	</Configurations>
@@ -234,6 +424,14 @@
 					/>
 				</FileConfiguration>
 				<FileConfiguration
+					Name="Debug|x64"
+					>
+					<Tool
+						Name="VCCLCompilerTool"
+						CompileAs="2"
+					/>
+				</FileConfiguration>
+				<FileConfiguration
 					Name="Release|Win32"
 					>
 					<Tool
@@ -241,6 +439,14 @@
 						CompileAs="2"
 					/>
 				</FileConfiguration>
+				<FileConfiguration
+					Name="Release|x64"
+					>
+					<Tool
+						Name="VCCLCompilerTool"
+						CompileAs="2"
+					/>
+				</FileConfiguration>
 			</File>
 			<Filter
 				Name="Common"
@@ -281,6 +487,14 @@
 						/>
 					</FileConfiguration>
 					<FileConfiguration
+						Name="Debug|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
+					<FileConfiguration
 						Name="Release|Win32"
 						>
 						<Tool
@@ -288,6 +502,14 @@
 							CompileAs="2"
 						/>
 					</FileConfiguration>
+					<FileConfiguration
+						Name="Release|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
 				</File>
 				<File
 					RelativePath="..\Common\Dlgcode.c"
@@ -301,6 +523,14 @@
 						/>
 					</FileConfiguration>
 					<FileConfiguration
+						Name="Debug|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
+					<FileConfiguration
 						Name="Release|Win32"
 						>
 						<Tool
@@ -308,6 +538,14 @@
 							CompileAs="2"
 						/>
 					</FileConfiguration>
+					<FileConfiguration
+						Name="Release|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
 				</File>
 				<File
 					RelativePath="..\Common\EncryptionThreadPool.c"
@@ -333,6 +571,14 @@
 						/>
 					</FileConfiguration>
 					<FileConfiguration
+						Name="Debug|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
+					<FileConfiguration
 						Name="Release|Win32"
 						>
 						<Tool
@@ -340,6 +586,14 @@
 							CompileAs="2"
 						/>
 					</FileConfiguration>
+					<FileConfiguration
+						Name="Release|x64"
+						>
+						<Tool
+							Name="VCCLCompilerTool"
+							CompileAs="2"
+						/>
+					</FileConfiguration>
 				</File>
 				<File
 					RelativePath="..\Common\Language.c"
@@ -374,6 +628,10 @@
 					>
 				</File>
 				<File
+					RelativePath="..\Common\Wipe.c"
+					>
+				</File>
+				<File
 					RelativePath="..\Common\Wipe.h"
 					>
 				</File>
@@ -523,6 +781,38 @@
 			UniqueIdentifier="{67DA6AB6-F800-4c08-8B7A-83BB121AAD01}"
 			>
 			<File
+				RelativePath="..\Boot\Windows\Rescue_Serpent_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Rescue_AES_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Rescue_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Release_Twofish_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Release_Serpent_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Release_AES_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Release_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
+				RelativePath="..\Boot\Windows\Rescue_Twofish_SHA2\BootSector.bin"
+				>
+			</File>
+			<File
 				RelativePath=".\Drive_icon_96dpi.bmp"
 				>
 			</File>
@@ -559,46 +849,46 @@
 				>
 			</File>
 			<File
-				RelativePath="..\Common\TrueCrypt_mounted.ico"
+				RelativePath="..\Common\VeraCrypt_mounted.ico"
 				>
 			</File>
 			<File
-				RelativePath="..\Common\TrueCrypt_volume.ico"
+				RelativePath="..\Common\VeraCrypt_volume.ico"
 				>
 			</File>
 			<Filter
 				Name="Common"
 				>
 				<File
-					RelativePath="..\Boot\Windows\Release_AES\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Rescue_Twofish\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Release\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Release_Serpent\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Release_Twofish\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Release_AES\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Rescue_Serpent\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Release\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Rescue_AES\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Release_Twofish\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Rescue\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Rescue_Serpent\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Rescue_Twofish\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Rescue_AES\BootLoader.com.gz"
 					>
 				</File>
 				<File
-					RelativePath="..\Boot\Windows\Release_Serpent\BootLoader.com.gz"
+					RelativePath="..\Boot\Windows\Rescue\BootLoader.com.gz"
 					>
 				</File>
 				<File
@@ -645,6 +935,14 @@
 						/>
 					</FileConfiguration>
 					<FileConfiguration
+						Name="Debug|x64"
+						ExcludedFromBuild="true"
+						>
+						<Tool
+							Name="VCResourceCompilerTool"
+						/>
+					</FileConfiguration>
+					<FileConfiguration
 						Name="Release|Win32"
 						ExcludedFromBuild="true"
 						>
@@ -652,6 +950,14 @@
 							Name="VCResourceCompilerTool"
 						/>
 					</FileConfiguration>
+					<FileConfiguration
+						Name="Release|x64"
+						ExcludedFromBuild="true"
+						>
+						<Tool
+							Name="VCResourceCompilerTool"
+						/>
+					</FileConfiguration>
 				</File>
 				<File
 					RelativePath="..\Boot\Windows\Release\Decompressor.com"
@@ -678,11 +984,43 @@
 					>
 				</File>
 				<File
-					RelativePath="..\Common\TrueCrypt.ico"
+					RelativePath="..\Common\VeraCrypt.ico"
 					>
 				</File>
 			</Filter>
 		</Filter>
+		<File
+			RelativePath="..\Boot\Windows\Rescue_AES_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Rescue_Serpent_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Rescue_Twofish_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Release_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Release_AES_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Release_Serpent_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Release_Twofish_SHA2\BootLoader.com.gz"
+			>
+		</File>
+		<File
+			RelativePath="..\Boot\Windows\Rescue_SHA2\BootLoader.com.gz"
+			>
+		</File>
 	</Files>
 	<Globals>
 	</Globals>
